nemo/collections/asr/parts/submodules/adapters/transformer_multi_head_attention_adapter_module.py

```diff
@@ -0,0 +1,128 @@
+    """
+
+    def __init__(
+        self,
+        hidden_size: int,
+        num_attention_heads: int,
+        attn_score_dropout: float = 0.0,
+        attn_layer_dropout: float = 0.0,
+        proj_dim: Optional[int] = None,
+        adapter_strategy: MHAResidualAddAdapterStrategy = None,
+    ):
+        super().__init__(
+            hidden_size=hidden_size,
+            num_attention_heads=num_attention_heads,
+            attn_score_dropout=attn_score_dropout,
+            attn_layer_dropout=attn_layer_dropout,
+        )
+
+        self.pre_norm = nn.LayerNorm(hidden_size)
+
+        # Set the projection dim to number of heads automatically
+        if proj_dim is not None and proj_dim < 1:
+            proj_dim = num_attention_heads
+
+        self.proj_dim = proj_dim
+
+        # Recompute weights for projection dim
+        if self.proj_dim is not None:
+            if self.proj_dim % num_attention_heads != 0:
+                raise ValueError(f"proj_dim ({proj_dim}) is not divisible by n_head ({num_attention_heads})")
+
+            self.attn_head_size = self.proj_dim // num_attention_heads
+            self.attn_scale = math.sqrt(math.sqrt(self.attn_head_size))
+            self.query_net = nn.Linear(hidden_size, self.proj_dim)
+            self.key_net = nn.Linear(hidden_size, self.proj_dim)
+            self.value_net = nn.Linear(hidden_size, self.proj_dim)
+            self.out_projection = nn.Linear(self.proj_dim, hidden_size)
+
+        # Setup adapter strategy
+        self.setup_adapter_strategy(adapter_strategy)
+
+        # reset parameters for Q to be identity operation
+        self.reset_parameters()
+
+    def forward(self, queries, keys, values, attention_mask):
+        """Compute 'Scaled Dot Product Attention'.
+        Args:
+            query (torch.Tensor): (batch, time1, size)
+            key (torch.Tensor): (batch, time2, size)
+            value(torch.Tensor): (batch, time2, size)
+            mask (torch.Tensor): (batch, time1, time2)
+            cache (torch.Tensor) : (batch, time_cache, size)
+
+        returns:
+            output (torch.Tensor): transformed `value` (batch, time1, d_model) weighted by the query dot key attention
+            cache  (torch.Tensor) : (batch, time_cache_next, size)
+        """
+        # Need to perform duplicate computations as at this point the tensors have been
+        # separated by the adapter forward
+        query = self.pre_norm(queries)
+        key = self.pre_norm(keys)
+        value = self.pre_norm(values)
+
+        return super().forward(query, key, value, attention_mask)
+
+    def reset_parameters(self):
+        with torch.no_grad():
+            nn.init.zeros_(self.out_projection.weight)
+            nn.init.zeros_(self.out_projection.bias)
+
+    def get_default_strategy_config(self) -> 'dataclass':
+        return MHAResidualAddAdapterStrategyConfig()
+
+
+@dataclass
+class TransformerMultiHeadAttentionAdapterConfig:
```

Comment:
minor: I would prefer MHA instead of long Multiheadattention name in class names.

Reply:
The config name matches the adapter class name + "Config", and the adapter class name is just "Adapter" at the end of the class that it subclasses (TransformerMultiHeadAttention). This conversion is carried over so we can always search for the parent class just by removing "Adapter" or "Config"